pkg/controller/container-runtime-config/helpers.go

```diff
@@ -209,39 +218,89 @@ func updateStorageConfig(data []byte, internal *mcfgv1.ContainerRuntimeConfigura
 	return newData.Bytes(), nil
 }
 
-// updateCRIOConfig decodes the data rendered from the template, merges the changes in and encodes it
-// back into a TOML format. It returns the bytes of the encoded data
-func updateCRIOConfig(data []byte, internal *mcfgv1.ContainerRuntimeConfiguration) ([]byte, error) {
-	tomlConf := new(tomlConfigCRIO)
-	if _, err := toml.DecodeReader(bytes.NewBuffer(data), tomlConf); err != nil {
-		return nil, fmt.Errorf("error decoding crio config: %v", err)
-	}
+// createCRIOdropinConfigLogLevel decodes the data rendered from the template, merges the changes in and encodes it
+// back into a TOML format. It returns the bytes of the encoded data.
+// This is only for when the log-level field is changed. We have separated it out into 3 different toml structs
+// and 3 different functions because the toml encoder does not ignore a value that was not set - it will set it as
+// empty. So if we just change field A and not B, it sets B to empty which is usually different from the default it
+// was set to.
+func createCRIOdropinConfigLogLevel(internal *mcfgv1.ContainerRuntimeConfiguration) ([]byte, error) {
```

Comment:
These three functions feel rather laborious.

How about a single
```go
func addTOMLFileConfig([]fileConfig, path string, interface{} marshalWithToml) ([]fileConfig, error)
```

called like
```go
	if ctrcfg.LogLevel != "" {
 		conf := tomlConfigCRIOLogLevel{}
                conf.Crio.Runtime.LogLevel = internal.LogLevel
 		fileConfigList, err := addTOMLFileConfig(fileConfigList, crioDropInFilePathLogLevel, conf)
 		if err { … }
      }
```

That would, AFAICS, make `createCRIODropinFiles` almost as large as the current implementation, but without splitting the logic over 4 functions and possibly worrying about whether the “if parameter is set” and “use parameter value” code in different functions can ever get out of sync, and without so much duplication.

Reply:
Agreed, fixed